src/main.rs

```diff
@@ -0,0 +1,433 @@
+use std::{hash::{DefaultHasher, Hash, Hasher}};
+
+//Hashvalues are of type u64. The type aliasing to HashValue is there to help with code readability.
+
+type HashValue = u64;
+
+/// Merkle tree struct definition
+/// Root is distinguished for easier access and code legibility
+/// layers represents the layers from the leaves up to the layer before the root, with the leaves array being the first and the topmost layer the last
+/// unique_elements represents the amount of non-repeated elements in the leaves
+struct MerkleTree {
+    root: HashValue,
+    layers: Vec<Vec<HashValue>>,
+    unique_elements: usize
+}
+
+/// Takes an element and returns its hash
+/// Input: a reference to a hashable element which we want the hash of
+/// Output: the element's hash
+fn hash<T: Hash> (val: &T) -> HashValue {
+    //For now I use the default hasher
+    //TODO: Hash algorithm to use could be made parametric
+    let mut hasher = DefaultHasher::new();
+    val.hash(&mut hasher);
+    let hash = hasher.finish();
+    hash
+}
+
+/// Calculates the closest power of 2 that's larger than the input
+/// Auxiliary function used for some operations over the tree
+fn closest_bigger_power_of2 (num: f64) -> u32 {
+    let exp = f64::log2(num).ceil();
+    f64::powf(2.0, exp) as u32
+}
+
+impl MerkleTree {
+    /// Merkle tree creation function
+    /// Input: A vector of data
+    /// Output: a Merkle tree with the hashes of said data as its leaves
+    pub fn new<T: Hash> (data: Vec<T>) -> MerkleTree {
+
+        // Hash every element
+        let mut hashes: Vec<HashValue> = data.iter().map(| elem | hash(elem)).collect();
+
+        // The amount of leaves always has to be a power of two so the tree remains balanced
+        // We achieve this by copying the last element the necessary amount of times to make the size of the leaves array a power of 23
```

Comment:
This comment should be like this, right?

```suggestion
        // We achieve this by copying the last element the necessary amount of times to make the size of the leaves array a power of 2
```

Reply:
Fixed it, thanks!